### ISSUE TITLE
Don't use 2 separate methods for sample/population

### ISSUE THREAD

**Issue by NathanJang** [evgenyneu/SigmaSwiftStatistics#2]
Have you considered using an enum or a boolean instead of 2 separate implementations for some methods?

For instance

`public static func variance(values: [Double], type: DataType) -> Double?` with `DataType.Sample` or something

or

`public static func variance(values: [Double], isSample: Bool) -> Double?`

and checking for it in the implementation instead of copy-pasting 2 very similar implementations.

**Comment by evgenyneu** [comment]
Good suggestion, thanks. I would not do it though, here are the reasons:

1. That would entangle two separate statistical concepts (sample and population) in the README. Currently they are completely separate entities, which is nice and simple for people to read and understand. Each of them have different formulas, return values and spreadsheet usage examples. 

2. The names of the methods will become less descriptive in my opinion: `Sigma.varianceSample([1, 12, 19.5, -5, 3, 8])` vs `Sigma.variance([1, 12, 19.5, -5, 3, 8], type: .Sample)`. In this case I like how the method name alone describes its nature.

3. This is true that sample and population methods contain some duplication, but they also have some subtle differences. For example, they can vary in the return values. Of course we can try removing the duplication completely and this is a viable option but creating a single method with one or two conditional statements. But in this case I prefer having some duplication in the methods that reduces their *individual* complexity. Duplication is bad, but only sometimes, depends on the context.

Thanks for your input. Have a good one!

**Comment by NathanJang** [comment]
:+1:

**Comment by NathanJang** [comment]
I just remembered another point that I forgot to bring up. I'm not quite sure your method names fit the naming conventions Apple uses. Personally I'd rename

`public static func varianceSample(values: [Double]) -> Double?`

to

`public static func varianceOfSample(values: [Double]) -> Double?`

or perhaps

`public static func variance(sample values: [Double]) -> Double?`

Up to you.

**Comment by Schemetrical** [comment]
`public static func variance(sample values: [Double]) -> Double?` goes well with `public static func variance(population values: [Double]) -> Double?` :smile:

**Comment by evgenyneu** [comment]
@NathanJang good point about naming, I need to think about it.

**Comment by evgenyneu** [comment]
@Schemetrical yes it is a good alternative, thank you. In this particular case I still prefer to keep information of "what the function does" in its name, and not to split it into the function name and parameter name. I described my motivations above.

**Comment by NathanJang** [comment]
`Sigma.variance(sample: [1, 12, 19.5, -5, 3, 8])` and `Sigma.variance(population: [1, 12, 19.5, -5, 3, 8])` would be 2 separate method signatures. Even though it's technically a parameter, it still distinguishes itself from the other.

**Comment by evgenyneu** [comment]
Now I am rethinking this and I am starting to like it. It does play beautifully with Swift parameters. The name of the parameter, population or sample, actually describes the data we pass into. And the parameter name still remains a part of the function name. Beautiful.

**Comment by Schemetrical** [comment]
Yep, and this is basically what Apple is doing while converting UIKit methods to swift with constructors too, doing `SomeObject(param1: a, param2: b)` has a different name than `SomeObject(otherParam1: a, otherParam2: b)` which is pretty neat.
